File: 3.无重复字符的最长子串.go
/*
 * @lc app=leetcode.cn id=3 lang=golang
 *
 * [3] 无重复字符的最长子串
 */
package leetcode

// @lc code=start
func lengthOfLongestSubstring(s string) int {
	ht := map[byte]int{}
	l, left, ans := len(s), 0, 0
	for i := 0; i < l; i++ {
		if v, ok := ht[s[i]]; ok {
			for j := left; j <= v; j++ {
				delete(ht, s[j])
			}
			left = v + 1
		}
		ht[s[i]] = i
		if i-left+1 > ans {
			ans = i - left + 1
		}
	}
	return ans
}

// @lc code=end

File: 3.无重复字符的最长子串.cpp
// @before-stub-for-debug-begin
#include <vector>
#include <string>
#include <unordered_set>
#include "commoncppproblem3.h"

using namespace std;
// @before-stub-for-debug-end

/*
 * @lc app=leetcode.cn id=3 lang=cpp
 *
 * [3] 无重复字符的最长子串
 */

// @lc code=start
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if(s.empty())   return 0;
        int ret = 0, right = -1;
        unordered_set<char> uomap;
        for (int left = 0; left < s.size(); left++) {
            if(left)
                uomap.erase(s[left - 1]);
            while(!uomap.count(s[right + 1]) && right + 1 < s.size()) {
                uomap.insert(s[right + 1]);
                right++;
            }
            ret = max(ret, right - left + 1);
        }
        
        return ret;
    }
};
// @lc code=end

